feat(OpenResponse): Add FeedbackRuleEvaluator expressions

hirokiterashima opened the pull request:
Part 1 of work for issue #829 "feat(OpenResponse): Add FeedbackRuleEvaluator". These are planned future PR's for this issue:
- refactoring, organizing folders to share common classes between OR and DG
- adding FeedbackRule authoring to OR

## Changes
- Evaluate FeedbackRules if authored for the OpenResponse content (see "Test Prep" section below). If the component content has cRater.feedbackRules, assume that it will use the FeedbackRuleEvaluator to provide feedback instead of other feedback mechanisms (e.g. MutlipleRoundsFeedback). Also assume that the author can specify one or more feedback texts for a FeedbackRule.
- Refactoring to share classes between DG and OR
   - Created abstract FeedbackRuleComponent that extends ComponentStudent and use this in DialogGuidanceFeedbackRuleEvaluator

## Test Prep
Using component advanced authoring, directly edit the component JSON and add "feedbackRules" block to OpenResponse's "cRater" block. Make sure that the feedback rules contain an id and a feedback array (instead of feedback string)
```
...
"cRater": {
  "feedback": {
    "enabled": true,
    "rules": [
      {
        "id": "lq3duxi3se",
        "expression": "ideaCountMoreThan(2)",
        "feedback": [
          "ideaCountMoreThan 2: first feedback",
          "ideaCountMoreThan 2: second feedback,
        ]
      },
      {
        "id": "default",
        "expression": "isDefault",
        "feedback": [
          "The default feedback"
        ]
      },
      ...
    ]
  }
  ...
}
...
```

## Test
- FeedbackRule evaluation for OR works and the correct feedback is shown to the student. 
   - If the rule has more than one feedback text, it should cycle through them.
- Existing C-Rater items that do not use FeedbackRule should work as before
- DG works as before

hirokiterashima committed:
feat(OpenResponse): Add FeedbackRuleEvaluator

#829

geoffreykwan reviewed (changes requested):
Functionality works well. I have two suggestions.

1. FeedbackRuleComponent is completely abstract so it can actually just be an interface. Then the OpenResponseStudent and DialogGuidanceStudent can implement that interface instead of extend it.
2. We have a boolean field enableMultipleAttemptScoringRules so should we also have a boolean field like enableFeedbackRules? This way authors can easily toggle it on and off without having to completely delete the feedbackRules array.

hirokiterashima committed:
Change FeedbackRules content to object instead of array #829

hirokiterashima committed:
Convert FeedbackRuleComponent to interface #829

hirokiterashima commented:
Good suggestions. I made the changes. 

Instead of enableFeedbackRules boolean, I converted FeedbackRules from Array to Object, and added a "enabled" and "rules" fields inside it. See updated descriptions to change the content.

hirokiterashima committed:
Change setting field name from "feedbackRules" to "feedback" #829

geoffreykwan reviewed (approved):
Looks good.
